backend/api/contacts/src/main/java/co/airy/core/contacts/ContactsController.java

```diff
@@ -0,0 +1,176 @@
+        }
+
+        if (contact == null) {
+            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(new RequestErrorResponsePayload("Contact not found"));
+        }
+
+        return ResponseEntity.ok(ContactResponsePayload.fromContact(contact));
+    }
+
+    @PostMapping("/contacts.update")
+    public ResponseEntity<?> updateContact(@RequestBody @Valid UpdateContactPayload payload) {
+        final String id = payload.getId().toString();
+        final Contact contact = stores.getContact(id);
+        if (contact == null) {
+            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(new RequestErrorResponsePayload("Contact not found"));
+        }
+
+        final Contact.Address address = Optional.ofNullable(contact.getAddress()).orElse(new Contact.Address());
+
+        final Contact updatedContact = contact.toBuilder()
+                .metadata(Optional.ofNullable(payload.getMetadata()).orElse(contact.getMetadata()))
+                .address(Optional.ofNullable(payload.getAddress())
+                        .map((payloadAddress) -> address.toBuilder()
+                                .addressLine1(Optional.ofNullable(payloadAddress.getAddressLine1()).orElse(address.getAddressLine1()))
+                                .addressLine2(Optional.ofNullable(payloadAddress.getAddressLine2()).orElse(address.getAddressLine2()))
+                                .city(Optional.ofNullable(payloadAddress.getCity()).orElse(address.getCity()))
+                                .country(Optional.ofNullable(payloadAddress.getCountry()).orElse(address.getCountry()))
+                                .postalCode(Optional.ofNullable(payloadAddress.getPostalCode()).orElse(address.getPostalCode()))
+                                .organizationName(Optional.ofNullable(payloadAddress.getOrganizationName()).orElse(address.getOrganizationName()))
+                                .build()
+                        )
+                        .orElse(address))
+                .conversations(Optional.ofNullable(payload.getConversations()).orElse(contact.getConversations()))
+                .displayName(Optional.ofNullable(payload.getDisplayName()).orElse(contact.getDisplayName()))
+                .avatarUrl(Optional.ofNullable(payload.getAvatarUrl()).orElse(contact.getAvatarUrl()))
+                .gender(Optional.ofNullable(payload.getGender()).orElse(contact.getGender()))
+                .locale(Optional.ofNullable(payload.getLocale()).orElse(contact.getLocale()))
+                .organizationName(Optional.ofNullable(payload.getOrganizationName()).orElse(contact.getOrganizationName()))
+                .timezone(Optional.ofNullable(payload.getTimezone()).orElse(contact.getTimezone()))
+                .title(Optional.ofNullable(payload.getTitle()).orElse(contact.getTitle()))
+                .via(Optional.ofNullable(payload.getVia()).orElse(contact.getVia()))
+                .build();
```

Comment:
It is a bit difficult to read what is going on here. maybe it can improve readability by moving the map part to a function like so
```java
final addressBuilder = (address, payloadAddress) -> {...}
```
And/Or have a helper function for: `Optional.ofNullable(payload.getConversations()).orElse(contact.getConversations())` 
So we cold do

```java
.conversations(get(payload.getConversations(), contact.getConversations()))
```

I don't know if this is even possible in java. 🙈

Reply:
I cleaned it up, by moving the address related merging code into the address class: https://github.com/airyhq/airy/pull/2628/files#diff-8aba54f3e719ce007647fefe1fdaddcf2df144d8fbe338a2a25e92ceab82a611R122